src/scala/io/bazel/rules_scala/jmh_support/BenchmarkGenerator.scala

```diff
@@ -131,17 +141,33 @@ object BenchmarkGenerator {
       tmpResourceDir.toFile.mkdir()
       tmpSourceDir.toFile.mkdir()
 
-      withClassLoader(benchmarkJarPath :: classpath) {
-        val source = new ASMGeneratorSource
-        val destination = new FileSystemDestination(tmpResourceDir.toFile, tmpSourceDir.toFile)
-        val generator = new JMHGenerator
-
-        collectClassesFromJar(benchmarkJarPath).foreach { path =>
-          // this would fail due to https://github.com/bazelbuild/rules_scala/issues/295
-          // let's throw a useful message instead
-          sys.error("jmh in rules_scala doesn't work with Java 8 bytecode: https://github.com/bazelbuild/rules_scala/issues/295")
-          source.processClass(Files.newInputStream(path))
+      withClassLoader(benchmarkJarPath :: classpath) { isolatedClassLoader =>
+
+        val source: GeneratorSource = generatorType match {
+          case AsmGenerator =>
+            val generatorSource = new ASMGeneratorSource
+            try {
+              generatorSource.processClasses(collectClassesFromJar(benchmarkJarPath).map(_.toFile).asJavaCollection)
+            } catch {
+              case _: ArrayIndexOutOfBoundsException =>
+                // this would fail due to https://github.com/bazelbuild/rules_scala/issues/295
+                // let's throw a useful message instead
+                sys.error("jmh in rules_scala doesn't work with Java 8 bytecode: https://github.com/bazelbuild/rules_scala/issues/295")
```

Comment:
can we change the error to say asm mode does not work with java 8 bytecode and instead suggest using reflection?